File: forssea_utilities/example/grafcet_code.py
#!/usr/bin/env python

from forssea_utilities.grafcet import GrafcetStep
from forssea_utilities.grafcet import GrafcetTransition
from forssea_utilities.grafcet import GrafcetCondition
from forssea_utilities.grafcet import Grafcet
from forssea_utilities.grafcet import Action

    
class DriverS0(Action):
    def __init__(self, grafcet):
        super(DriverS0, self).__init__(grafcet)
        self.initialize()

    def execute(self):
        # To be implemented
        pass

                    
class DriverS0S1Condition(GrafcetCondition):
    def __init__(self, grafcet):
        super(DriverS0S1Condition, self).__init__("DriverS0S1", grafcet)

    def eval(self):
        # To be implemented
        return False

                    
class DriverS1(Action):
    def __init__(self, grafcet):
        super(DriverS1, self).__init__(grafcet)
        self.initialize()

    def execute(self):
        # To be implemented
        pass

                    
class DriverS1S2Condition(GrafcetCondition):
    def __init__(self, grafcet):
        super(DriverS1S2Condition, self).__init__("DriverS1S2", grafcet)

    def eval(self):
        # To be implemented
        return False

                    
class DriverS2(Action):
    def __init__(self, grafcet):
        super(DriverS2, self).__init__(grafcet)
        self.initialize()

    def execute(self):
        # To be implemented
        pass

                    
class DriverS2S3Condition(GrafcetCondition):
    def __init__(self, grafcet):
        super(DriverS2S3Condition, self).__init__("DriverS2S3", grafcet)

    def eval(self):
        # To be implemented
        return False

                    
class DriverS3(Action):
    def __init__(self, grafcet):
        super(DriverS3, self).__init__(grafcet)
        self.initialize()

    def execute(self):
        # To be implemented
        pass

                    
class DriverS3S4Condition(GrafcetCondition):
    def __init__(self, grafcet):
        super(DriverS3S4Condition, self).__init__("DriverS3S4", grafcet)

    def eval(self):
        # To be implemented
        return False

                    
class DriverS4(Action):
    def __init__(self, grafcet):
        super(DriverS4, self).__init__(grafcet)
        self.initialize()

    def execute(self):
        # To be implemented
        pass

                    
class DriverS4S3Condition(GrafcetCondition):
    def __init__(self, grafcet):
        super(DriverS4S3Condition, self).__init__("DriverS4S3", grafcet)

    def eval(self):
        # To be implemented
        return False

                    
class DriverS1S0Condition(GrafcetCondition):
    def __init__(self, grafcet):
        super(DriverS1S0Condition, self).__init__("DriverS1S0", grafcet)

    def eval(self):
        # To be implemented
        return False

                    
class DriverS3S30Condition(GrafcetCondition):
    def __init__(self, grafcet):
        super(DriverS3S30Condition, self).__init__("DriverS3S30", grafcet)

    def eval(self):
        # To be implemented
        return False

                    
class DriverS30(Action):
    def __init__(self, grafcet):
        super(DriverS30, self).__init__(grafcet)
        self.initialize()

    def execute(self):
        # To be implemented
        pass

                    
class DriverS30S3Condition(GrafcetCondition):
    def __init__(self, grafcet):
        super(DriverS30S3Condition, self).__init__("DriverS30S3", grafcet)

    def eval(self):
        # To be implemented
        return False

                    
class DriverS30S300Condition(GrafcetCondition):
    def __init__(self, grafcet):
        super(DriverS30S300Condition, self).__init__("DriverS30S300", grafcet)

    def eval(self):
        # To be implemented
        return False

                    
class DriverS300(Action):
    def __init__(self, grafcet):
        super(DriverS300, self).__init__(grafcet)
        self.initialize()

    def execute(self):
        # To be implemented
        pass

                    
class DriverS3S300Condition(GrafcetCondition):
    def __init__(self, grafcet):
        super(DriverS3S300Condition, self).__init__("DriverS3S300", grafcet)

    def eval(self):
        # To be implemented
        return False

                    
class DriverS300S0Condition(GrafcetCondition):
    def __init__(self, grafcet):
        super(DriverS300S0Condition, self).__init__("DriverS300S0", grafcet)

    def eval(self):
        # To be implemented
        return False

                    
class DriverS4S40Condition(GrafcetCondition):
    def __init__(self, grafcet):
        super(DriverS4S40Condition, self).__init__("DriverS4S40", grafcet)

    def eval(self):
        # To be implemented
        return False

                    
class DriverS40(Action):
    def __init__(self, grafcet):
        super(DriverS40, self).__init__(grafcet)
        self.initialize()

    def execute(self):
        # To be implemented
        pass

                    
class DriverS40S3Condition(GrafcetCondition):
    def __init__(self, grafcet):
        super(DriverS40S3Condition, self).__init__("DriverS40S3", grafcet)

    def eval(self):
        # To be implemented
        return False

                    
class DriverS40S300Condition(GrafcetCondition):
    def __init__(self, grafcet):
        super(DriverS40S300Condition, self).__init__("DriverS40S300", grafcet)

    def eval(self):
        # To be implemented
        return False

                    
class DriverS4S41Condition(GrafcetCondition):
    def __init__(self, grafcet):
        super(DriverS4S41Condition, self).__init__("DriverS4S41", grafcet)

    def eval(self):
        # To be implemented
        return False

                    
class DriverS41(Action):
    def __init__(self, grafcet):
        super(DriverS41, self).__init__(grafcet)
        self.initialize()

    def execute(self):
        # To be implemented
        pass

                    
class DriverS41S3Condition(GrafcetCondition):
    def __init__(self, grafcet):
        super(DriverS41S3Condition, self).__init__("DriverS41S3", grafcet)

    def eval(self):
        # To be implemented
        return False

                    
class DriverS41S300Condition(GrafcetCondition):
    def __init__(self, grafcet):
        super(DriverS41S300Condition, self).__init__("DriverS41S300", grafcet)

    def eval(self):
        # To be implemented
        return False

                    
class DriverGrafcet(Grafcet):
    def __init__(self, frequency):
        super(DriverGrafcet, self).__init__("DriverGrafcet", frequency)

    def _specific_setup(self):

        # Step 0
        transitions = []
        action = DriverS0(self)
                    
        # Transition S0 / S1 Condition
        upstream_step_ids = [0]
        downstream_step_ids = [1]
        transitions.append(GrafcetTransition(DriverS0S1Condition(self), upstream_step_ids, downstream_step_ids))
                                
        step_object = GrafcetStep(0, transitions, action)
        step_object.set_active(True)
        self._add_step(step_object)

        # Step 1
        transitions = []
        action = DriverS1(self)
                    
        # Transition S1 / S2 Condition
        upstream_step_ids = [1]
        downstream_step_ids = [2]
        transitions.append(GrafcetTransition(DriverS1S2Condition(self), upstream_step_ids, downstream_step_ids))
                                
        # Transition S1 / S0 Condition
        upstream_step_ids = [1]
        downstream_step_ids = [0]
        transitions.append(GrafcetTransition(DriverS1S0Condition(self), upstream_step_ids, downstream_step_ids))
                                
        step_object = GrafcetStep(1, transitions, action)
        step_object.set_active(False)
        self._add_step(step_object)

        # Step 2
        transitions = []
        action = DriverS2(self)
                    
        # Transition S2 / S3 Condition
        upstream_step_ids = [2]
        downstream_step_ids = [3]
        transitions.append(GrafcetTransition(DriverS2S3Condition(self), upstream_step_ids, downstream_step_ids))
                                
        step_object = GrafcetStep(2, transitions, action)
        step_object.set_active(False)
        self._add_step(step_object)

        # Step 3
        transitions = []
        action = DriverS3(self)
                    
        # Transition S3 / S4 Condition
        upstream_step_ids = [3]
        downstream_step_ids = [4]
        transitions.append(GrafcetTransition(DriverS3S4Condition(self), upstream_step_ids, downstream_step_ids))
                                
        # Transition S3 / S30 Condition
        upstream_step_ids = [3]
        downstream_step_ids = [30]
        transitions.append(GrafcetTransition(DriverS3S30Condition(self), upstream_step_ids, downstream_step_ids))
                                
        # Transition S3 / S300 Condition
        upstream_step_ids = [3]
        downstream_step_ids = [300]
        transitions.append(GrafcetTransition(DriverS3S300Condition(self), upstream_step_ids, downstream_step_ids))
                                
        step_object = GrafcetStep(3, transitions, action)
        step_object.set_active(False)
        self._add_step(step_object)

        # Step 4
        transitions = []
        action = DriverS4(self)
                    
        # Transition S4 / S3 Condition
        upstream_step_ids = [4]
        downstream_step_ids = [3]
        transitions.append(GrafcetTransition(DriverS4S3Condition(self), upstream_step_ids, downstream_step_ids))
                                
        # Transition S4 / S40 Condition
        upstream_step_ids = [4]
        downstream_step_ids = [40]
        transitions.append(GrafcetTransition(DriverS4S40Condition(self), upstream_step_ids, downstream_step_ids))
                                
        # Transition S4 / S41 Condition
        upstream_step_ids = [4]
        downstream_step_ids = [41]
        transitions.append(GrafcetTransition(DriverS4S41Condition(self), upstream_step_ids, downstream_step_ids))
                                
        step_object = GrafcetStep(4, transitions, action)
        step_object.set_active(False)
        self._add_step(step_object)

        # Step 30
        transitions = []
        action = DriverS30(self)
                    
        # Transition S30 / S3 Condition
        upstream_step_ids = [30]
        downstream_step_ids = [3]
        transitions.append(GrafcetTransition(DriverS30S3Condition(self), upstream_step_ids, downstream_step_ids))
                                
        # Transition S30 / S300 Condition
        upstream_step_ids = [30]
        downstream_step_ids = [300]
        transitions.append(GrafcetTransition(DriverS30S300Condition(self), upstream_step_ids, downstream_step_ids))
                                
        step_object = GrafcetStep(30, transitions, action)
        step_object.set_active(False)
        self._add_step(step_object)

        # Step 300
        transitions = []
        action = DriverS300(self)
                    
        # Transition S300 / S0 Condition
        upstream_step_ids = [300]
        downstream_step_ids = [0]
        transitions.append(GrafcetTransition(DriverS300S0Condition(self), upstream_step_ids, downstream_step_ids))
                                
        step_object = GrafcetStep(300, transitions, action)
        step_object.set_active(False)
        self._add_step(step_object)

        # Step 40
        transitions = []
        action = DriverS40(self)
                    
        # Transition S40 / S3 Condition
        upstream_step_ids = [40]
        downstream_step_ids = [3]
        transitions.append(GrafcetTransition(DriverS40S3Condition(self), upstream_step_ids, downstream_step_ids))
                                
        # Transition S40 / S300 Condition
        upstream_step_ids = [40]
        downstream_step_ids = [300]
        transitions.append(GrafcetTransition(DriverS40S300Condition(self), upstream_step_ids, downstream_step_ids))
                                
        step_object = GrafcetStep(40, transitions, action)
        step_object.set_active(False)
        self._add_step(step_object)

        # Step 41
        transitions = []
        action = DriverS41(self)
                    
        # Transition S41 / S3 Condition
        upstream_step_ids = [41]
        downstream_step_ids = [3]
        transitions.append(GrafcetTransition(DriverS41S3Condition(self), upstream_step_ids, downstream_step_ids))
                                
        # Transition S41 / S300 Condition
        upstream_step_ids = [41]
        downstream_step_ids = [300]
        transitions.append(GrafcetTransition(DriverS41S300Condition(self), upstream_step_ids, downstream_step_ids))
                                
        step_object = GrafcetStep(41, transitions, action)
        step_object.set_active(False)
        self._add_step(step_object)

File: forssea_utilities/example/grafcet_code.cpp

#include <grafcet_code.hpp>
    
bool DriverS0::execute()
{
    return true;
}
                    
bool DriverS0S1Condition::eval()
{
    return false;
}
                    
bool DriverS1::execute()
{
    return true;
}
                    
bool DriverS1S2Condition::eval()
{
    return false;
}
                    
bool DriverS2::execute()
{
    return true;
}
                    
bool DriverS2S3Condition::eval()
{
    return false;
}
                    
bool DriverS3::execute()
{
    return true;
}
                    
bool DriverS3S4Condition::eval()
{
    return false;
}
                    
bool DriverS4::execute()
{
    return true;
}
                    
bool DriverS4S3Condition::eval()
{
    return false;
}
                    
bool DriverS1S0Condition::eval()
{
    return false;
}
                    
bool DriverS3S30Condition::eval()
{
    return false;
}
                    
bool DriverS30::execute()
{
    return true;
}
                    
bool DriverS30S3Condition::eval()
{
    return false;
}
                    
bool DriverS30S300Condition::eval()
{
    return false;
}
                    
bool DriverS300::execute()
{
    return true;
}
                    
bool DriverS3S300Condition::eval()
{
    return false;
}
                    
bool DriverS300S0Condition::eval()
{
    return false;
}
                    
bool DriverS4S40Condition::eval()
{
    return false;
}
                    
bool DriverS40::execute()
{
    return true;
}
                    
bool DriverS40S3Condition::eval()
{
    return false;
}
                    
bool DriverS40S300Condition::eval()
{
    return false;
}
                    
bool DriverS4S41Condition::eval()
{
    return false;
}
                    
bool DriverS41::execute()
{
    return true;
}
                    
bool DriverS41S3Condition::eval()
{
    return false;
}
                    
bool DriverS41S300Condition::eval()
{
    return false;
}
                    
DriverGrafcetGrafcet::DriverGrafcetGrafcet(ros::NodeHandle* node_handle, const float& frequency)
: Grafcet(string("DriverGrafcetGrafcet"), node_handle, frequency)
{
    this->setup();
}

void DriverGrafcetGrafcet::specificSetup()
{
    list<GrafcetTransition*> transitions;
    list<int> upstream_step_ids;
    list<int> downstream_step_ids;
    Action* action = 0;
    GrafcetStep* step = 0;

    // Step 0
    transitions.clear();
    action = new DriverS0(this);
                    
    // Tansition S0 / S1 Condition
    upstream_step_ids.clear();
    downstream_step_ids.clear();
    upstream_step_ids.push_back(0);
    downstream_step_ids.push_back(1);
    transitions.push_back(new GrafcetTransition(new DriverS0S1Condition(), upstream_step_ids, downstream_step_ids));
                                
    step = new GrafcetStep(0, transitions, action);
    step->setActive(true);
    this->addStep(step);
    // Step 1
    transitions.clear();
    action = new DriverS1(this);
                    
    // Tansition S1 / S2 Condition
    upstream_step_ids.clear();
    downstream_step_ids.clear();
    upstream_step_ids.push_back(1);
    downstream_step_ids.push_back(2);
    transitions.push_back(new GrafcetTransition(new DriverS1S2Condition(), upstream_step_ids, downstream_step_ids));
                                
    // Tansition S1 / S0 Condition
    upstream_step_ids.clear();
    downstream_step_ids.clear();
    upstream_step_ids.push_back(1);
    downstream_step_ids.push_back(0);
    transitions.push_back(new GrafcetTransition(new DriverS1S0Condition(), upstream_step_ids, downstream_step_ids));
                                
    step = new GrafcetStep(1, transitions, action);
    step->setActive(false);
    this->addStep(step);
    // Step 2
    transitions.clear();
    action = new DriverS2(this);
                    
    // Tansition S2 / S3 Condition
    upstream_step_ids.clear();
    downstream_step_ids.clear();
    upstream_step_ids.push_back(2);
    downstream_step_ids.push_back(3);
    transitions.push_back(new GrafcetTransition(new DriverS2S3Condition(), upstream_step_ids, downstream_step_ids));
                                
    step = new GrafcetStep(2, transitions, action);
    step->setActive(false);
    this->addStep(step);
    // Step 3
    transitions.clear();
    action = new DriverS3(this);
                    
    // Tansition S3 / S4 Condition
    upstream_step_ids.clear();
    downstream_step_ids.clear();
    upstream_step_ids.push_back(3);
    downstream_step_ids.push_back(4);
    transitions.push_back(new GrafcetTransition(new DriverS3S4Condition(), upstream_step_ids, downstream_step_ids));
                                
    // Tansition S3 / S30 Condition
    upstream_step_ids.clear();
    downstream_step_ids.clear();
    upstream_step_ids.push_back(3);
    downstream_step_ids.push_back(30);
    transitions.push_back(new GrafcetTransition(new DriverS3S30Condition(), upstream_step_ids, downstream_step_ids));
                                
    // Tansition S3 / S300 Condition
    upstream_step_ids.clear();
    downstream_step_ids.clear();
    upstream_step_ids.push_back(3);
    downstream_step_ids.push_back(300);
    transitions.push_back(new GrafcetTransition(new DriverS3S300Condition(), upstream_step_ids, downstream_step_ids));
                                
    step = new GrafcetStep(3, transitions, action);
    step->setActive(false);
    this->addStep(step);
    // Step 4
    transitions.clear();
    action = new DriverS4(this);
                    
    // Tansition S4 / S3 Condition
    upstream_step_ids.clear();
    downstream_step_ids.clear();
    upstream_step_ids.push_back(4);
    downstream_step_ids.push_back(3);
    transitions.push_back(new GrafcetTransition(new DriverS4S3Condition(), upstream_step_ids, downstream_step_ids));
                                
    // Tansition S4 / S40 Condition
    upstream_step_ids.clear();
    downstream_step_ids.clear();
    upstream_step_ids.push_back(4);
    downstream_step_ids.push_back(40);
    transitions.push_back(new GrafcetTransition(new DriverS4S40Condition(), upstream_step_ids, downstream_step_ids));
                                
    // Tansition S4 / S41 Condition
    upstream_step_ids.clear();
    downstream_step_ids.clear();
    upstream_step_ids.push_back(4);
    downstream_step_ids.push_back(41);
    transitions.push_back(new GrafcetTransition(new DriverS4S41Condition(), upstream_step_ids, downstream_step_ids));
                                
    step = new GrafcetStep(4, transitions, action);
    step->setActive(false);
    this->addStep(step);
    // Step 30
    transitions.clear();
    action = new DriverS30(this);
                    
    // Tansition S30 / S3 Condition
    upstream_step_ids.clear();
    downstream_step_ids.clear();
    upstream_step_ids.push_back(30);
    downstream_step_ids.push_back(3);
    transitions.push_back(new GrafcetTransition(new DriverS30S3Condition(), upstream_step_ids, downstream_step_ids));
                                
    // Tansition S30 / S300 Condition
    upstream_step_ids.clear();
    downstream_step_ids.clear();
    upstream_step_ids.push_back(30);
    downstream_step_ids.push_back(300);
    transitions.push_back(new GrafcetTransition(new DriverS30S300Condition(), upstream_step_ids, downstream_step_ids));
                                
    step = new GrafcetStep(30, transitions, action);
    step->setActive(false);
    this->addStep(step);
    // Step 300
    transitions.clear();
    action = new DriverS300(this);
                    
    // Tansition S300 / S0 Condition
    upstream_step_ids.clear();
    downstream_step_ids.clear();
    upstream_step_ids.push_back(300);
    downstream_step_ids.push_back(0);
    transitions.push_back(new GrafcetTransition(new DriverS300S0Condition(), upstream_step_ids, downstream_step_ids));
                                
    step = new GrafcetStep(300, transitions, action);
    step->setActive(false);
    this->addStep(step);
    // Step 40
    transitions.clear();
    action = new DriverS40(this);
                    
    // Tansition S40 / S3 Condition
    upstream_step_ids.clear();
    downstream_step_ids.clear();
    upstream_step_ids.push_back(40);
    downstream_step_ids.push_back(3);
    transitions.push_back(new GrafcetTransition(new DriverS40S3Condition(), upstream_step_ids, downstream_step_ids));
                                
    // Tansition S40 / S300 Condition
    upstream_step_ids.clear();
    downstream_step_ids.clear();
    upstream_step_ids.push_back(40);
    downstream_step_ids.push_back(300);
    transitions.push_back(new GrafcetTransition(new DriverS40S300Condition(), upstream_step_ids, downstream_step_ids));
                                
    step = new GrafcetStep(40, transitions, action);
    step->setActive(false);
    this->addStep(step);
    // Step 41
    transitions.clear();
    action = new DriverS41(this);
                    
    // Tansition S41 / S3 Condition
    upstream_step_ids.clear();
    downstream_step_ids.clear();
    upstream_step_ids.push_back(41);
    downstream_step_ids.push_back(3);
    transitions.push_back(new GrafcetTransition(new DriverS41S3Condition(), upstream_step_ids, downstream_step_ids));
                                
    // Tansition S41 / S300 Condition
    upstream_step_ids.clear();
    downstream_step_ids.clear();
    upstream_step_ids.push_back(41);
    downstream_step_ids.push_back(300);
    transitions.push_back(new GrafcetTransition(new DriverS41S300Condition(), upstream_step_ids, downstream_step_ids));
                                
    step = new GrafcetStep(41, transitions, action);
    step->setActive(false);
    this->addStep(step);
}
